wandb/artifacts.py

```diff
@@ -1038,9 +1041,58 @@ def _extra_from_obj(obj):
             'versionID': obj.generation,
         }
 
-    @staticmethod
-    def _content_addressed_path(md5):
-        # TODO: is this the structure we want? not at all human
-        # readable, but that's probably OK. don't want people
-        # poking around in the bucket
-        return 'wandb/%s' % base64.b64encode(md5.encode("ascii")).decode("ascii")
+
+class HTTPHandler(StorageHandler):
+    def __init__(self, session, scheme=None):
+        self._scheme = scheme or "http"
+        self._cache = artifacts_cache.get_artifacts_cache()
+        self._session = session
+
+    @property
+    def scheme(self):
+        return self._scheme
+
+    def load_path(self, artifact, manifest_entry, local=False):
+        if not local:
+            return manifest_entry.ref
+
+        path, hit = self._cache.check_etag_obj_path(manifest_entry.digest, manifest_entry.size)
+        if hit:
+            return path
+
+        response = self._session.get(manifest_entry.ref, stream=True)
+        response.raise_for_status()
+
+        digest, size, extra = self._entry_from_headers(response.headers)
+        if manifest_entry.digest != digest:
+            raise ValueError('Digest mismatch for url %s: expected %s but found %s' %
+                             (manifest_entry.ref, manifest_entry.digest, digest))
+
+        with open(path, "wb") as file:
+            for data in response.iter_content(chunk_size=16 * 1024):
+                file.write(data)
+        return path
+
+    def store_path(self, artifact, path, name=None, checksum=True, max_objects=None):
+        name = name or os.path.basename(path)
+        if not checksum:
+            return [ArtifactManifestEntry(name, path, digest=path)]
+
+        with self._session.get(path, stream=True) as response:
```

Comment:
Nice, I wonder if there are cases where we would want to just do a `head` request here...

Reply:
With `stream=True` I think it behaves the same way as the requests library won't start pulling in the bytes until `response.content` is accessed, but I can make this an explicit `HEAD` request instead.

Reply:
I think it's better to use `head()`, will fix.

Reply:
Yeah, this is more robust.  Not all endpoints will support a HEAD request, but there could be some that only return these more detailed headers on a HEAD.  For now lets just keep it and maybe add a comment, I would think most will have the headers in the GET.